### ISSUE TITLE
MySqlTransaction.Dispose(Boolean disposing) throws System.InvalidOperationException: Connection must be Open; current state is Closed

### ISSUE THREAD

**Issue by jemiller0** [mysql-net/MySqlConnector#383]
I'm running into a problem where a query times out. Then, Dispose() gets called on the Transaction that is being used. I looked at the MySqlConnector code and it tries to rollback the transaction inside the Dispose() method. But, it appears that the connection is already closed. So, the Command that is trying to rollback the transaction throws an exception. This masks the original exception.

```
System.InvalidOperationException: Connection must be Open; current state is Closed
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQuery()
   at MySql.Data.MySqlClient.MySqlTransaction.Dispose(Boolean disposing)
   at OleLibrary.OleDapperContext.Dispose() in C:\Users\jemiller\Documents\Visual Studio 2017\Projects\Ole\OleLibrary\OleDapperContext.cs:line 2558

```

Would it be possible to modify the following code to include a check for whether the connection is open?

```
		protected override void Dispose(bool disposing)
		{
			try
			{
				if (disposing)
				{
					if (!m_isFinished && m_connection != null && m_connection.CurrentTransaction == this)
					{
						using (var cmd = new MySqlCommand("rollback", m_connection, this))
							cmd.ExecuteNonQuery();
						m_connection.CurrentTransaction = null;
					}
					m_connection = null;
				}
			}
			finally
			{
				base.Dispose(disposing);
			}
		}
```

**Comment by bgrainger** [comment]
Yes, that code should definitely be changed to not throw from `Dispose`; I'll push a fix.

**Comment by jemiller0** [comment]
Thanks. I think fixing that will help users of the library. I think I've seen that before. I don't remember if it was with your driver or another one. I think it may have been Oracle's driver. I would receive an error about a connection being closed and kept thinking something was malfunctioning on the server and the connection was being closed or something. But, in reality, I think it was a command timeout or other error happened, then another error happened in a Dispose() method and masked the original exception. I thought I read in the IDisposable documentation that one should never throw an exception in the Dispose() method. However, I looked and it just mentioned having the method called multiple times should always succeed.

**Comment by bgrainger** [comment]
Fixed in [0.29.4](https://github.com/mysql-net/MySqlConnector/releases/tag/0.29.4).

**Comment by jemiller0** [comment]
Thanks Bradley. You do great work. I was amazed at how quickly you fixed the last issue I reported. I am becoming more and more sold on GitHub and NuGet even though at times it can be frustrating dealing with DLL versioning issues. It is pretty awesome how quick you can push out a fix.
